### PR TITLE
napari test requirement

### PR DESCRIPTION
**Pull request by joshmoore**
Registering napari-conftest as a plugin in setup.py
makes a hard-requirement on napari. This PR moves the
requirement to tests only and registers the pytest
plugin at runtime.

With thanks to @Czaki

see #https://github.com/napari/napari/issues/1661

Note: that the environment.yml file is currently not tested by this PR and I in fact needed to actively remove `tox-conda` to get things to work. We will likely need to either remove it (in favor of an ome-zarr package on conda?) or add in separate testing.

### CODE REVIEW

**Comment by joshmoore** [comment]
@tlambert03 : somewhere you had pointed me to a Qt-testing-in-GitHub-Actions solution that you had come up with. Can you point me to that pointer?

**Comment by tlambert03** [comment]
here are a couple:
https://github.com/tlambert03/napari/blob/feature/github-tests/.github/workflows/tests.yml
https://github.com/napari/magicgui/blob/master/.github/workflows/test_and_deploy.yml

**Comment by Czaki** [comment]
I also have one https://github.com/4DNucleome/PartSeg/blob/master/.github/workflows/tests.yml

**Comment by joshmoore** [comment]
Import suggestion from @Czaki in https://github.com/ome/ome-zarr-py/pull/67 solved the issue, though I think it seems un-pytest-plugin-y. Now hitting a missing `on_draw` though it looks like the napari test suite still uses this.

```
299 def test_viewer(self, make_test_viewer):  # noqa
300 """example of testing the viewer."""
301 viewer = make_test_viewer()
302
303 shapes = [(4000, 3000), (2000, 1500), (1000, 750), (500, 375)]
304 np.random.seed(0)
305 data = [np.random.random(s) for s in shapes]
306 _ = viewer.add_image(data, multiscale=True, contrast_limits=[0, 1])
307 layer = viewer.layers[0]
308
309 # Set canvas size to target amount
310 viewer.window.qt_viewer.view.canvas.size = (800, 600)
311 >       viewer.window.qt_viewer.on_draw(None)
312 E       AttributeError: 'QtViewer' object has no attribute 'on_draw'
```

**Comment by Czaki** [comment]
This error with `on_draw` is connected with an older version of napari. This test should be disabled for python 3.6 (or has some inner if, to use different methods for this python version)

**Comment by Czaki** [comment]
I have already seen a problem with python 3.7 Pyside2 on Linux. I check in my code and skip this case in tests.

**Comment by joshmoore** [comment]
Thanks, @Czaki. Yeah, don't think I'll try to tackle this segfault:
```
=============================== warnings summary ===============================
.tox/py37-PySide2/lib/python3.7/site-packages/napari/utils/events/dataclass.py:184: 40 warnings
  /home/runner/work/ome-zarr-py/ome-zarr-py/.tox/py37-PySide2/lib/python3.7/site-packages/napari/utils/events/dataclass.py:184: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
    different = np.any(before != after)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
======================= 24 passed, 40 warnings in 11.55s =======================
Fatal Python error: Segmentation fault

Thread 0x00007f7735ffb700 (most recent call first):
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/multiprocessing/pool.py", line 470 in _handle_results
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 870 in run
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 926 in _bootstrap_inner
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 890 in _bootstrap

Thread 0x00007f77367fc700 (most recent call first):
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/multiprocessing/pool.py", line 422 in _handle_tasks
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 870 in run
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 926 in _bootstrap_inner
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 890 in _bootstrap

Thread 0x00007f7736ffd700 (most recent call first):
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/multiprocessing/pool.py", line 413 in _handle_workers
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 870 in run
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 926 in _bootstrap_inner
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 890 in _bootstrap

Thread 0x00007f77377fe700 (most recent call first):
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/multiprocessing/pool.py", line 110 in worker
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 870 in run
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 926 in _bootstrap_inner
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 890 in _bootstrap

Thread 0x00007f7737fff700 (most recent call first):
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/multiprocessing/pool.py", line 110 in worker
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 870 in run
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 926 in _bootstrap_inner
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 890 in _bootstrap

Thread 0x00007f777666f700 (most recent call first):
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 296 in wait
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 552 in wait
  File "/home/runner/work/ome-zarr-py/ome-zarr-py/.tox/py37-PySide2/lib/python3.7/site-packages/napari/components/experimental/chunk/_delay_queue.py", line 151 in run
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 926 in _bootstrap_inner
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/threading.py", line 890 in _bootstrap

Current thread 0x00007f778129b680 (most recent call first):
ERROR: InvocationError for command /home/runner/work/ome-zarr-py/ome-zarr-py/.tox/py37-PySide2/bin/pytest tests (exited with code -11 (SIGSEGV)) (exited with code -11)
```

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  test:
+    name: Test ${{ matrix.os }} / ${{ matrix.toxenv }}
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+
+          - {os: windows-latest, python_Version: '3.6', toxenv: 'py36-PyQt5'}
+          - {os: windows-latest, python_Version: '3.6', toxenv: 'py36-PySide2'}
```
Comment (Czaki) [inline .github/workflows/build.yml:15]: I see no py36-PyQt5 and py36-PySide2 entries in tox files.

File: tox.ini [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+# passenv = DISPLAY XAUTHORITY
+passenv = GITHUB_ACTIONS
+deps =
+    -rrequirements/requirements-dev.txt
+    pytest-xvfb ; sys_platform == 'linux'
+    PyQt5: PyQt5!=5.15.0
+    PySide2: PySide2!=5.15.0
+
+commands =
+    pip install .[napari]
```
Comment (Czaki) [inline tox.ini:19]: it is not necsssary

File: tox.ini [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+[testenv]
+# passenv = DISPLAY XAUTHORITY
+passenv = GITHUB_ACTIONS
+deps =
+    -rrequirements/requirements-dev.txt
+    pytest-xvfb ; sys_platform == 'linux'
+    PyQt5: PyQt5!=5.15.0
+    PySide2: PySide2!=5.15.0
+
```
Comment (Czaki) [inline tox.ini:17]:
```suggestion

extras = napari

```

**Review by Czaki — CHANGES_REQUESTED** [review]
Few remarks.

**Comment by joshmoore** [comment]
Thanks, @Czaki, but still getting failures on linux && (mac/py.39)

**Comment by Czaki** [comment]
> Thanks, @Czaki, but still getting failures on linux && (mac/py.39)

As I see it fails on installed `numcodecs` which dose not provide a wheel for python 3.9. 

```
**     c-blosc/internal-complibs/zlib-1.2.8/gzlib.c:256:24: error: implicit declaration of function 'lseek' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
**
```

maybe disable python 3.9 test with napari.

**Comment by Czaki** [comment]
It is hard to tell something about linux when see this: 
Zrzut ekranu z 2020-12-04 13-46-54

**Comment by joshmoore** [comment]
Agreed! That has only started happening today. As for numcodecs & the wheel, yes, the other alternative would be to go back to conda. What do you think?

**Comment by Czaki** [comment]
> Agreed! That has only started happening today. As for numcodecs & the wheel, yes, the other alternative would be to go back to conda. What do you think?

The big problem with conda is that it goes with preinstalled qt in the old version (5.9) which breaks napari tests. 

Personally, I will skip testing this for python 3.9 and open issue on `numcodecs` repository. Napari itself has no test for python 3.9. So creating test set which depends on untested napari may be not the best idea. 

If I good remember there is problem with vispy which install on python 3.9 but will not run (maybe it was solved? @sofoniew @jni did you know?)

**Comment by joshmoore** [comment]
> The big problem with conda is that it goes with preinstalled qt in the old version (5.9) which breaks napari tests.

Understood. Thanks for the heads up. Unfortunate then that we tried to do all of our workshops for I2K with conda.

> Napari itself has no test for python 3.9. So creating test set which depends on untested napari may be not the best idea.

Wow. Ok. I recently had issues using anything _other than_ 3.9 (at least on Mac Big Sur). I guess  3.8 is the sweet spot then.

**Comment by tlambert03** [comment]
> If I good remember there is problem with vispy which install on python 3.9 but will not run (maybe it was solved? @sofroniewn @jni did you know?)

that was solved in https://github.com/vispy/vispy/pull/1914 and released in [vispy 0.6.6](https://pypi.org/project/vispy/)

**Comment by Czaki** [comment]
> Wow. Ok. I recently had issues using anything _other than_ 3.9 (at least on Mac Big Sur). I guess 3.8 is the sweet spot then.

There is work on enabling python 3.8 on ARM BigSure. https://github.com/pypa/wheel/issues/387#issuecomment-738329574


> Understood. Thanks for the heads up. Unfortunate then that we tried to do all of our workshops for I2K with conda.

It does not break napari itself but breaks napari tests. Also It is possible to uninstall but it increases the complication of the process. Maybe it is simpler than I think. I'm not very familiar with conda, so maybe I chose the wrong way. Maybe someone from napari maintainers could help here.

**Comment by joshmoore** [comment]
Ok. As it stands, we're green. In summary:
 - Python 3.6 is only tested on Windows
 - Python 3.9 is not tested on Mac
 - *No* Python versions are tested on Linux
 - The `make_test_viewer` plugin still requires an odd import & registration (without registration `E           ValueError: no option named '--show-viewer' is thrown)

I think it's still enough testing for the purposes of ome-zarr, but help finding reliable CI testing across the spectrum would be appreciated, and I'd certainly also be interested in doing that with conda (both within CI and outwith)

**Comment by joshmoore** [comment]
P.S.

> It does not break napari itself but breaks napari tests.

We did have workshop attendees who had trouble with napari+conda.

File: requirements/requirements-dev.txt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+black
+napari
```
Comment (joshmoore) [inline requirements/requirements-dev.txt:2]: @Czaki, for a follow-up, thoughts on removing this if we're handling it via the "extras = " setting?

**Comment by joshmoore** [comment]
Merging. I'll be opening an immediate follow-up to activate coverage testing. Happy to make adjustments there.

**Pull request merged**